scripts/ioc_deploy.py

```diff
@@ -301,6 +426,143 @@ def make_in(deploy_dir: str, dry_run: bool) -> int:
         return subprocess.run(["make"], cwd=deploy_dir).returncode
 
 
+def set_permissions(deploy_dir: str, allow_write: bool, dry_run: bool) -> int:
+    """
+    Apply or remove write protection from a deploy repo.
+
+    You may or may not have permissions to do this to releases created by other users.
+
+    allow_write=True involves adding "w" permissions
+    for the owner and for the group. "w" permissions will never be added for other users.
+    We will also add write permissions to the top-level directory.
+
+    allow_write=False involves removing the "w" permissions
+    from all files, and from directories that contain untracked files.
+    We will also remove permissions from the top-level direcotry.
+    """
+    if dry_run and not os.path.isdir(deploy_dir):
+        # Dry run has nothing to do if we didn't build the dir
+        # Most things past this point will error out
+        logger.info("Dry-run: skipping permission changes on never-made directory")
+        return ReturnCode.SUCCESS
+
+    # Ignore these directories
+    exclude = (".git",)
+
+    if allow_write:
+        # Lazy and simple: chmod everything
+        perms = get_allow_write_rule(deploy_dir)
+        if dry_run:
+            logger.info(f"Dry-run: skipping chmod({deploy_dir}, {oct(perms)})")
+        else:
+            os.chmod(deploy_dir, perms)
+        for dirpath, dirnames, filenames in exclude_walk(deploy_dir, exclude=exclude):
+            for name in dirnames + filenames:
+                full_path = os.path.join(dirpath, name)
+                perms = get_allow_write_rule(full_path)
+                if dry_run:
+                    logger.debug(f"Dry-run: skipping chmod({full_path}, {oct(perms)})")
+                else:
+                    logger.debug(f"chmod({full_path}, {oct(perms)})")
+                    os.chmod(full_path, perms)
+        return ReturnCode.SUCCESS
+
+    # Compare the files that exist to the files that are tracked by git
+    try:
+        ls_files_output = subprocess.check_output(
+            ["git", "-C", deploy_dir, "ls-files"],
+            universal_newlines=True,
+        )
+    except subprocess.CalledProcessError as exc:
+        return exc.returncode
+    deploy_path = Path(deploy_dir)
+    tracked_paths = [
+        deploy_path / subpath.strip() for subpath in ls_files_output.splitlines()
+    ]
+    build_dir_paths = set()
+
+    def accumulate_subpaths(subpath: Path):
```

Comment:
Is there a reason this is defined within the scope of `set_permissions` rather than just being a standalone function that takes `build_dir_paths`,  `tracked_paths`, `exclude`, and `tracked_paths` as arguments?  

I like to avoid this sort of closure-type scoping but it's not strictly wrong.  It does save a good amount of typing

Reply:
I'm using the closure scoping here as a lazy way to mutate my set, there's no technical reason or logic flow why it has to stay this way. It could equivalently be a standalone function or pair of functions that create and return sets, or it could be a function I pass a set into to be mutated.

Reply:
I'm happy to change it for better code quality